Add support for tokens inside ramp and animation

DRKblade opened the pull request:
This is a continuation of issue #2223.

```
[module/battery]
type = internal/battery
battery = BAT1
adapter = ADP1

format-charging = <animation-charging>
format-charging-background = #000
format-charging-foreground = #fff

format-discharging = <ramp-capacity>
format-discharging-background = #fff
format-discharging-foreground = #000
format-discharging-padding = 1

ramp-capacity = %label% %percentage%%
ramp-capacity-0 = 
ramp-capacity-1 = 
ramp-capacity-2 = 

animation-charging = %label% %percentage%%
animation-charging-0 = 
animation-charging-1 = 
animation-charging-2 = 

format-full = <label-full>
format-full-background = #000
format-full-foreground = #fff
format-full-padding = 1
label-full =  %percentage%% 
```

patrick96 commented:
Does this also include all the commits from #2199?

Also, do I understand this correctly:

```dosini
ramp-capacity = %label% %percentage%%
ramp-capacity-0 = 
ramp-capacity-1 = 
ramp-capacity-2 = 
```

`%label%` is replaced with the contents of the ramp entry that is currently active?

Would the following also be possible?

```dosini
ramp-capacity-0 =  %percentage%% 
ramp-capacity-1 =  %percentage%%
ramp-capacity-2 =  %percentage%%
```

DRKblade commented:
You understood my idea correctly.

This pull request accidentally include all the commits from #2199. I implemented this on top of the commits from #2199 since token support should be provided to the ramp and animation of warn states too. I hope #2199 is accepted.

Supporting tokens in individual icons or frames is totally possible, I just didn't think it was necessary since I thought they were only used to display warnings, and I have already implemented warn states by that point.

patrick96 commented:
> Supporting tokens in individual icons or frames is totally possible, I just didn't think it was necessary since I thought they were only used to display warnings, and I have already implemented warn states by that point.

I think this is something that we should support. It would make ramps and animations much more flexible.

> This pull request accidentally include all the commits from #2199. I implemented this on top of the commits from #2199 since token support should be provided to the ramp and animation of warn states too. I hope #2199 is accepted.

Alright, I will review #2199 first then.

patrick96 commented:
@DRKblade Please ping me once this is ready to review again.

DRKblade commented:
The branch is now mergable, I'm working on the individual icons/frames label.

DRKblade commented:
@patrick96 I finished it, you can review the code now.

patrick96 reviewed (changes requested):
Code looks good. I like the `labellist` abstraction :)

This no longer includes support for the root property and the `%label%` token for ramps and animations, correct?

The following modules also have ramps or animations which are not implemented by this PR:

* alsa
* backlight
* network
* pulseaudio
* xbacklight

Also, please add your changes to the `CHANGELOG.md` file: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog

DRKblade commented on src/modules/battery.cpp:
```diff
@@ -247,6 +235,20 @@ namespace modules {
+      replace_tokens(m_label_full);
+    } else if (m_state == battery_module::state::DISCHARGING) {
+      replace_tokens(m_label_discharging);
+      replace_tokens(m_animation_discharging);
```
It took me some digging to discover that :)

patrick96 replied:
I would really love to just use C++20, but most distros don't support it yet :frowning_face:

DRKblade commented on tests/common/config_test.hpp:
```diff
@@ -0,0 +1,22 @@
+  return label->get();
+}
+
+config::make_type load_config(std::string&& path, std::string&& bar) {
```
I don't use it anymore, I'm gonna delete it.

DRKblade commented on src/modules/cpu.cpp:
```diff
@@ -156,6 +168,33 @@ namespace modules {
 
+  /**
+   * Subthread runner that emits update events to refresh animations
+   * in case they are used. Note, that it is ok to
+   * use a single thread, because the two animations are never shown at the
+   * same time.
```
Oops, that's why you don't copy and paste in programming.

DRKblade commented:
I can't figure out the build errors, why the doesn't compiler build `labellist.cpp`, only to complain that the methods in it are missing. I built successfully on my machine.

patrick96 commented:
> I can't figure out the build errors, why the doesn't compiler build src/labellist.cpp, only to complain that the methods in it are missing. I built successfully on my machine.

On `master` you need to add each new source file to `src/CMakeLists.txt`. This is quite new, it's possible that you don't have that yet in your branch. But I think CI does a merge first and there obivously, this is a requirement.

DRKblade commented:
Ok, I think the branch is mergeable now.

patrick96 commented on CHANGELOG.md:
```diff
@@ -8,6 +8,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+- Support for tokens inside ramp and animation. 
+  ([`#2223`](https://github.com/polybar/polybar/issues/2223), 
+  [`#682`](https://github.com/polybar/polybar/issues/682)) 
+  
```
Please also add a separate entry for the cpu warning animation.

prabh1601 commented:
Uh guys, can you please resolve this pull request? I am literally waiting for this feature to get merged

patrick96 commented:
@prabh1601 This gets done when it gets done. No need to rush anyone.
